### ISSUE TITLE
Stripe outage question

### ISSUE THREAD

**Issue by chubarovNick** [andreoji/naive_dice#3]
What if stripe has no outage it’s network issue when the app does not receive success charge response it sounds like code charge user and not approve his reservation because of rollback.

**Comment by andreoji** [comment]
Hello again Nick,
I think this scenario is covered by all the `error` responses of the `stripity_stripe` client.
I believe the network error scenario is covered by these API test cases in the stripe code base itself L35 - L50:
https://github.com/code-corps/stripity_stripe/blob/4c26d86b2d3869a0b361205cbad34cdb8696b1db/test/stripe/api_test.exs#L35
So the client will respond to the app with one of the error tuples in the scenario you're describing.
As a result the transaction will roll back.
I may have misunderstood your concern, if so shout.
I think you've alluded to a lack of test coverage around the `batches` module itself.

**Comment by chubarovNick** [comment]
I mean case when stripe finish successfully, but stripe client does not receive a response because of a network outage. It means that we rollback transaction but charge money from user

**Comment by andreoji** [comment]
The Stripe API supports idempotency for safely retrying requests without accidentally performing the same operation twice.

So in the scenario described - if an api request to create a charge does not respond (but does actually creating the charge) due to a network connection error, you can retry the request with the same idempotency key to guarantee that no more than one charge is created.

Subsequent requests will return the same result as the first request, in this scenario we would get back the original charge object.

`stripity_stripe` sets the idempotency key by default, here is the request you made today:
```
ID
req_MEr7GhInoWa7tI
Time
17/01/2020, 14:10:06
Method
POST
URL
/v1/charges
Status
200
IP address
54.205.64.231
Version
2019-05-16
Source
Stripe/v1 stripity-stripe/2019-05-16
Related
charge — ch_1G1vWUHCCcwyjBBX4gHdxOBS
Idempotency
Key — 2nlb3okjfkgk00f44o0000g2
```
Request retries can be configured in `stripity_stripe`
To set retries, you can pass the number of attempts and range of backoff (time between attempting the request again) in milliseconds.

`config :stripity_stripe, :retries, [max_attempts: 3, base_backoff: 500, max_backoff: 2_000]`

This retry configuration isn't in place currently as you know.

The refund endpoint could be on possible enhancement once the number of retries had been exhausted. But that in itself would be dependent on the quality of the network connection.

So in short to answer your question we would need to retry the charge endpoint using the idempotency key and retries mechanism within the context of the transaction.

**Comment by vovayartsev** [comment]
Thanks, Andre, it's a valid point about Idempotency Key, and retries will decrease the risk almost to a zero. And still, if we want to build a 100% robust system that can work without manual interventions - what are our options here? 

You mentioned one option already - scheduling a call to "refund endpoint" so that the money don't remain locked on the user's account. Can you please explain in more detail how it works (taking into consideration that we don't know charge_id if the charge request failed)?

What are the other options with Stripe to make sure we don't steal the money from a user without issuing a ticket?

**Comment by andreoji** [comment]
Hello Vladimir
Thanks for your input.

To have a 100% robust ticketing system we would have to have a single "source of truth" in terms of what has actually been paid for. This would be the state of Stripe system itself, and the charges created within - ultimately a charge is either successful or it isn't.

The state of the Stripe system is exposed by the Events endpoint. Every change of state in Stripe, generates an Event i.e., the creation of a charge, the creation of a refund etc.

If we create a charge, but don't get an actual response back, we can subsequently query the Events endpoint after the timeout period. In this way, we are always checking the actual state of Stripe in terms of our inputs and expected outcomes.

You are correct the charge_id would be required to refund, and wouldn't be directly available in our scenario, due to network issues. 

One solution to finding the charge_id would be the Events endpoint:

Retrieve the list of recent Events, then find the Event of interest - searching on whatever parameters yield the correct Event or an empty result - the `created` timestamp, `name` and `type` being options.
e.g.
```
"created": 1579270206
"name": "paul@acme.com"
"type": "charge.succeeded"
```
Once we have the Event object its charge_id can be used to call the Refund endpoint.

Other options within Stripe to make sure we don't steal money from the user would be Payment Intents

The PaymentIntent objects act as the single source of truth in the lifecycle of the flow.

A PaymentIntent with a status of succeeded means that the payment flow it is driving is complete.

The funds are now in the account, we can confidently fulfill the order and issue the ticket.

If we need to refund the customer, again we can use the Refunds API.
